### ISSUE TITLE
RuntimeError(f'Store label {label} is not a string; provide a label_encoder to StoreConfig')

### ISSUE THREAD

**Issue by daniel-a-diaz** [static-frame/static-frame#605]
I am trying to export a yarn to_zip_npz, but I am getting this runtime error. How exactly do I fix this? For instance here is a toy example that kindly follows my use case. How would I make this example work?

    from pathlib import Path
    import numpy as np
    import static_frame as sf

    path = Path(f'datasets')
    npz_file = path / 'npz_yarn_test.zip'

    frame_data = np.array([[1,2,3],[4,5,6],[7,8,9]])
    frames = []
    for i in range(1,4):
        frame = sf.Frame(frame_data, columns=('first', 'second', 'third'), name = f'frame_{i}')
        frames.append(frame)
    
    buses = []
    for i in range(1,4):
        bus = sf.Bus.from_frames(frames, name = f'bus_{i}')
        buses.append(bus)
    
    sf.Yarn.from_buses(buses, retain_labels=True).to_zip_npz(npz_file)

**Comment by flexatone** [comment]
Many thanks, @daniel-a-diaz , for your question and clear example code.

The issue here is that the `Yarn` you have created has an `IndexHierarchy` since you set `retain_labels=True`:

```python
>>> yn
<Yarn>
<IndexHierarchy>
bus_1            frame_1 Frame
bus_1            frame_2 Frame
bus_1            frame_3 Frame
bus_2            frame_1 Frame
bus_2            frame_2 Frame
bus_2            frame_3 Frame
bus_3            frame_1 Frame
bus_3            frame_2 Frame
bus_3            frame_3 Frame
<<U5>            <<U7>   <object>
```

When we go to write to an NPZ, we use the `Yarn` labels to name the components in the file store, and those components must be strings. The labels given from an `IndexHierarchy`, however, are tuples.  We can encode and decode those tuples using a `StoreConfig`, providing a function to encode (going to a string) and decode (going from string back to the label). Thus, to round-trip this `Yarn` to and from a ZIP NPZ, we can do the following:

```
>>> yn.to_zip_npz('/tmp/npz.zip', config=sf.StoreConfig(label_encoder=str))

>>> import ast
>>> sf.Bus.from_zip_npz('/tmp/npz.zip', config=sf.StoreConfig(label_decoder=ast.literal_eval), index_constructor=sf.IndexHierarchy.from_labels)
<Bus>
<IndexHierarchy>
bus_1            frame_1 <FrameDeferred>
bus_1            frame_2 <FrameDeferred>
bus_1            frame_3 <FrameDeferred>
bus_2            frame_1 <FrameDeferred>
bus_2            frame_2 <FrameDeferred>
bus_2            frame_3 <FrameDeferred>
bus_3            frame_1 <FrameDeferred>
bus_3            frame_2 <FrameDeferred>
bus_3            frame_3 <FrameDeferred>
<<U5>            <<U7>   <object>
```
Note that while the `label_decoder` will give us back tuples, we still need to provide an `index_constructor` to create an `IndexHierachy`.

Please let me know if you have further questions.

**Comment by daniel-a-diaz** [comment]
So I ran into a problem with the export
     sf.Yarn.from_buses(buses, retain_labels = True).to_zip_npz(self.filename, config = sf.StoreConfig(label_encoder=str))

It said: No support for object dtypes. I figured out that it was because I had dictionaries and tuples in my frames, and those are considered objects. Is there anyway to construct a frame so that it accepts dictionaries and tuples as a dtype?

**Comment by flexatone** [comment]
Yes, NumPy object arrays cannot be written to NPZ, as NPZ is stable binary format for C-typed data. If you want to store object arrays to disk, the best option is to use `to_zip_pickle()` / `from_zip_pickle()`, though there are some considerations to keep in mind when using pickles.

https://docs.python.org/3/library/pickle.html

**Comment by daniel-a-diaz** [comment]
Is there a way to do something similar as you did with the labels, where I encode an entire column as str() for export, and then after import, when I access an entry from that column just decode it with ast.literal.eval(). I am trying to use f.assign['column'].apply(lambda) but this isn't working. Even when I try the example from the docs, it isn't working. Like I can't access any attributes, methods, or anything from the created frame.

    f = sf.Frame.from_fields(((10, -2, 0, 0), (8, -3, 8, 0), (1, 0, 9, 12)), index=('p', 'q', 'r', 's'), columns=('a', 'b', 'c'), name='x')
    f.assign['a'].apply(lambda s: s / 100)

**Comment by daniel-a-diaz** [comment]
So I ended up just dealing with converting my dictionaries and tuples to str() earlier in my code and it works fine. But now I am struggling with how to actually access my frames. Below is another toy example.

    import static_frame as sf
    from pathlib import Path
    import ast
    
    path = Path(f'datasets')
    npz_file = path / 'npz_yarn_test.zip'

    frame_data = [(10, str({'key': 0}), str((0,1,2))), (10, str({'key': 0}), str((0,1,2))), (10, str({'key': 0}), str((0,1,2)))]

    frames = []
    for i in range(1,4):
        frame = sf.Frame.from_records(frame_data, columns=('first', 'second', 'third'), name = f'frame_{i}')
        frames.append(frame)

    buses = []
    for i in range(1,4):
        bus = sf.Bus.from_frames(frames, name = f'bus_{i}')
        buses.append(bus)

    sf.Yarn.from_buses(buses, retain_labels=True).to_zip_npz(npz_file, config=sf.StoreConfig(label_encoder=str))

    bus = sf.Bus.from_zip_npz(npz_file, config=sf.StoreConfig(label_decoder=ast.literal_eval), index_constructor=sf.IndexHierarchy.from_labels)

It looks like the keys are tuples. But if I try something like
`print(bus[('bus_1', 'frame_1')])`
I get the error: Store label ('bus_1', 'frame_1') is not a string; provide a label_encoder to StoreConfig

So I tried 
`print(bus[str(('bus_1', 'frame_1'))])`
But get this error: Invalid key length for ('bus_1', 'frame_1'); must be length 2.

From here I've tried all kinds of things but get KeyErrors. So I guess I have two questions. How do I actually access my frames, and what might be the best practices for dealing with this kind of setup?

And thank you so much for all the help, and the incredibly prompt replies. I really appreciate it.

**Comment by flexatone** [comment]
Hi @daniel-a-diaz , thanks for you for your persistence.

The issue you are having is that the `Bus`, once loaded from the ZIP NPZ, continues to need to be able to translate back and forth between the string labels on disk and the in memory labels. Thus, you need to create your `Bus` with both a  `label_decoder` and a `label_encoder`. Then it works as expected:

```python
>>> bus = sf.Bus.from_zip_npz('/tmp/bus.zip', config=sf.StoreConfig(label_decoder=ast.literal_eval, label_encoder=str), index_constructor=sf.IndexHierarchy.from_labels)                                                                                                                                      
>>> bus[('bus_1', 'frame_1')]                                                                                                                            
<Frame: frame_1>
<Index>          int     dict       tuple     <<U5>
<Index>
0                10      {'key': 0} (0, 1, 2)
1                10      {'key': 0} (0, 1, 2)
2                10      {'key': 0} (0, 1, 2)
<int64>          <int64> <<U10>     <<U9>
```
Regarding managing your data, I can offer that putting containers (dicts, tuples) inside of a `Frame` is a little unconventional. Better would be to flatten-out your data so that, for example, each key in your dictionary becomes its own column. You might also benefit from not trying to put all the data in on table: maybe one table for metadata, and another for NumPy data. Happy to help further if you want to elaborate on your data layout.

**Comment by daniel-a-diaz** [comment]
My data consists of CT scans that are segmented into several different classes. From these segmentations I am extracting all the individual volumes along with their associated labels and metadata, and feeding them into my network. Much like you suggested I have been keeping the actual 3D numpy volumes separate from the metadata that I have in a frame. For the numpy volumes I am using `savez()` and `savez_compressed()` where I have a npz for each class and dataset type (training, validation, test) full of individual volume arrays. My volumes are all resized to the exact same size and are segmentations so it's all binary. I had one smaller file compress from 15.8 GB to 56 MB, so I'm hoping that is a sign loading compressed npzs might not take too big of a hit when loading volumes. Keeping things compressed would be ideal unless it is too big of a hit. Each volume entry in a frame has a npz file location, array name, class label, dataset type, and other metadata. This is the way I had been setting of my Dataset to put in the DataLoader.
     
    class MyDataset(Dataset):        
        def __init__(self, df, transform = None):
            self.npzs = df['npz']
            self.array_names = df['array_name']
            self.labels = df['label']
            self.transform = transform

        def __getitem__(self, index):
            with ZipFile(self.npzs[index], 'r') as npz_file:
                np_vol = np.load(npz_file.open(self.array_name[index]))
                torch_vol = torch.Tensor(np_vol).unsqueeze(0).repeat(3,1,1,1)
            if self.transform: # TODO: get transorms figured out
                torch_vol = self.transform(torch_vol)
            label = self.labels[index]
            return torch_vol, label

        def __len__(self):
            return len(self.filenames)

For the frames and exporting, the kind of hierarchy I am envisioning is to have a Yarn consisting of a group of CT scans I'm wanting to investigate, where each CT scan is a bus containing a frame for each class I am segmenting out of the scan. And for importing my thought is to make a Quilt from the Yarn I saved, and then make new Quilts out of that, filtering based on the dataset type. So each type of DataLoader (train, validation, test) is using one of these new Quilts of the same type to reference. And then follow much the same as the code above for creating the Datasets for the DataLoader.

**Comment by flexatone** [comment]
Apologies for not getting back to your questions! I hope this is still helpful:

1. NPZ files writen by NumPy are not directly readable by StaticFrame. Using the context manager of `ZipFile` only loads what you request, so it is lazy.
2. Yes, using `to_zip_npz` on a `Bus` is the fastest stable file format, and includes all characteristics of each `Frame`. You can use pickles for more speed, but they are not good for long-term storage (and NPZ is just as fast in some scenarios.
3. Yes, an example will help. But in short, tuples in an `IndexHierarchy` should work like any other label.
4. With a `Qult` loaded from a `Bus` configured to read from a file store, lazy loading is implicit and will always happen. What determines this is the `Bus` that you give to the `Quilt`.

Apologies again for my delay. Please feel free to post more questions or issues if the come up.
